// swaggiffy/build/compiled/helpers/registerDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerDefinition = void 0;
const globals_1 = require("../globals");
/**
 * Create swagger path definition
 * @param router Express router
 * @returns apiPathDefinitions {APIPathDefinition}
 */
function registerDefinition(router, options) {
    const paths = router.stack.filter((item) => item.route);
    paths.forEach((item) => {
        var _a;
        const method = item.route.stack[0].method.toLowerCase();
        const path = item.route.path;
        let parameters = [];
        let responses = {};
        if (item.keys.length > 0) {
            for (const key of item.keys) {
                parameters.push({
                    in: 'path',
                    name: key.name,
                    required: true,
                    type: 'string',
                });
            }
        }
        if (method == 'post' || method == 'put') {
            parameters.push({
                in: 'body',
                name: 'body',
                required: method == 'post' ? true : false,
                schema: {
                    $ref: `#/definitions/${options.mappedSchema}`,
                },
            });
        }
        if (method == 'delete') {
            responses = {
                '200': {
                    description: 'OK',
                    schema: {
                        type: 'object',
                        properties: {
                            deleted: {
                                type: 'boolean',
                                example: true,
                            },
                        },
                    },
                },
                '500': {
                    description: 'Internal Server Error',
                },
            };
        }
        else if (method == 'post' || method == 'put') {
            responses = {
                '201': {
                    description: 'Created',
                    schema: {
                        $ref: `#/definitions/${options.mappedSchema}`,
                    },
                },
                '500': {
                    description: 'Internal Server Error',
                },
            };
        }
        else {
            responses = {
                '200': {
                    description: 'OK',
                    schema: {
                        type: 'array',
                        items: {
                            $ref: `#/definitions/${options.mappedSchema}`,
                        },
                    },
                },
                '500': {
                    description: 'Internal Server Error',
                    schema: {
                        type: 'object',
                        properties: {
                            error: {
                                type: 'string',
                                example: 'Internal Server Error',
                            },
                        },
                    },
                },
            };
        }
        const pathDefinition = {
            pathString: `${options.basePath}${path}`,
            tags: ((_a = options.tags) === null || _a === void 0 ? void 0 : _a.split(' ')) || [],
            method: method,
            meta: {
                summary: options.summary || '',
                description: options.description || '',
                // operationId: `${method}`,
                parameters: parameters,
                produces: options.produces || ['application/json'],
                consumes: options.consumes || ['application/json'],
                responses,
                security: [
                    {
                        Bearer: ['global'],
                    },
                ],
            },
        };
        (0, globals_1.getAPIDefinitionMetadataStorage)().apiDefinitions.push({
            router,
            apiDefinition: pathDefinition,
        });
    });
}
exports.registerDefinition = registerDefinition;
// export function registerDefinitions(registerMetas: APIRegisterMeta[]) {
//     registerMetas.forEach((_meta) => {
//         registerDefinition(_meta);
//     });
// }
