// swaggiffy/build/compiled/commands/GenerateConfigCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateConfigCommand = void 0;
const PlatformTools_1 = require("../platform/PlatformTools");
const SetupRunner_1 = require("../runners/SetupRunner");
/**
 * Generate Config Command
 */
class GenerateConfigCommand {
    constructor() {
        this.command = 'generate:config';
        this.describe = 'Generate swaggiffy config file.';
        this.aliases = 'g:config';
    }
    builder(args) {
        return args
            .option('path', {
            alias: 'configFilePath',
            type: 'string',
            describe: 'File where the config file should be created. Defaults to BASE_DIR/swaggiffy.config.json .',
        })
            .option('r', {
            alias: 'refresh',
            type: 'boolean',
            describe: 'Re-generate and overwrite existing config file.',
        });
    }
    async handler(args) {
        try {
            const override = args.refresh ? true : false;
            const configFile = await SetupRunner_1.SetupRunner.generateConfigFile(GenerateConfigCommand.getOSA2Template(), override);
            PlatformTools_1.PlatformTools.logSuccess('Successfully generated config file');
        }
        catch (err) {
            PlatformTools_1.PlatformTools.logCmdErr('Error when generating config file: ', err);
        }
    }
    /**
     * Generate Config 0SA2 template
     * @param projectName Project Name
     * @returns template
     */
    static getOSA2Template(projectName) {
        return JSON.stringify({
            projectName: projectName || 'swaggiffy',
            openApiVersion: '2.0',
            outFile: 'src/swagger.json',
            appPort: 5008,
            apiBaseUrl: '/api',
            swaggerRoute: '/api-docs',
        }, undefined, 3);
    }
    /**
     * Generate Config 0SA3 template
     * @param projectName Project Name
     * @returns template
     */
    static getOSA3Template(projectName) {
        return JSON.stringify({
            projectName: projectName || 'swaggiffy',
            openApiVersion: '3.0',
            outFile: 'src/swagger.json',
            appPort: 5008,
            apiBaseUrl: '/api',
            swaggerRoute: '/api-docs',
        }, undefined, 3);
    }
}
exports.GenerateConfigCommand = GenerateConfigCommand;
