// swaggiffy/build/compiled/utils/Templates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Templates = void 0;
const PlatformTools_1 = require("../platform/PlatformTools");
const Defaults_1 = require("./Defaults");
/**
 * Swaggiffy Templates class
 */
class Templates {
    /**
     * Returns swaggiffy config file template
     * @param options
     * @returns
     */
    static getConfigTemplate(options) {
        return JSON.stringify({
            projectName: (options === null || options === void 0 ? void 0 : options.projectName) || PlatformTools_1.PlatformTools.getProjectName(),
            openApiVersion: (options === null || options === void 0 ? void 0 : options.openApiVersion) || Defaults_1.Defaults.OPENAPI_VERSION,
            outFile: (options === null || options === void 0 ? void 0 : options.outFile) || Defaults_1.Defaults.SWAGGER_DEFINITION_FILE,
            apiRoute: (options === null || options === void 0 ? void 0 : options.apiRouteUrl) || Defaults_1.Defaults.SWAGGER_ENDPOINT_URL,
            format: (options === null || options === void 0 ? void 0 : options.format) || Defaults_1.Defaults.SWAGGER_DEFINITION_FORMAT,
        }, undefined, 3);
    }
    /**
     * Generate Config 0SA2 template
     * @param projectName Project Name
     * @returns template
     */
    static getOSA2Template(projectName, port) {
        const name = projectName ? projectName : PlatformTools_1.PlatformTools.getProjectName();
        return JSON.stringify({
            swaggerDefinition: {
                swagger: '2.0',
                info: {
                    title: name,
                    description: `${name} API Documentation`,
                    termsOfService: 'http://swagger.io/terms/',
                    contact: {
                        name: 'API Support',
                        url: 'http://www.swagger.io/support',
                        email: 'support@swagger.io',
                    },
                    license: {
                        name: 'Apache 2.0',
                        url: 'http://www.apache.org/licenses/LICENSE-2.0.html',
                    },
                    version: '1.0.0',
                },
                host: `localhost:${port || '5008'}`,
                basePath: '/',
                schemes: ['http'],
                securityDefinitions: {
                    Bearer: {
                        type: 'apiKey',
                        name: 'Authorization',
                        in: 'header',
                    },
                },
                paths: {},
            },
            apis: [],
        }, undefined, 3);
    }
    /**
     * Generate Config 0SA3 template
     * @param projectName Project Name
     * @returns template
     */
    static getOSA3Template(projectName, port) {
        const name = projectName ? projectName : PlatformTools_1.PlatformTools.getProjectName();
        return JSON.stringify({
            swaggerDefinition: {
                openapi: '3.0.0',
                info: {
                    title: name,
                    description: `${name} API Documentation`,
                    termsOfService: 'http://swagger.io/terms/',
                    contact: {
                        name: 'API Support',
                        url: 'http://www.swagger.io/support',
                        email: 'support@swagger.io',
                    },
                    license: {
                        name: 'Apache 2.0',
                        url: 'http://www.apache.org/licenses/LICENSE-2.0.html',
                    },
                    version: '1.0.0',
                },
                host: `localhost:${port || '5008'}`,
                basePath: '/',
                schemes: ['http'],
                paths: {},
                components: {
                    securitySchemes: {
                        bearerAuth: {
                            type: 'http',
                            scheme: 'bearer',
                            bearerFormat: 'JWT',
                        },
                    },
                },
            },
            apis: [],
        }, undefined, 3);
    }
}
exports.Templates = Templates;
